src/VisualStudio/Xaml/Impl/Implementation/LanguageServer/Handler/Diagnostics/AbstractPullDiagnosticHandler.cs

```diff
@@ -0,0 +1,182 @@
+                var previousResultId = documentToPreviousResultId.ContainsKey(document) ? documentToPreviousResultId[document] : null;
+
+                // Call XamlPullDiagnosticService to get the diagnostic report for this document.
+                // We will compute what to report inside XamlPullDiagnosticService, for example, whether we should keep using the previousId or use a new resultId,
+                // and the handler here just return the result get from XamlPullDiagnosticService.
+                var diagnosticReport = await _xamlDiagnosticService.GetDiagnosticReportAsync(document, previousResultId, cancellationToken).ConfigureAwait(false);
+                progress.Report(CreateReport(
+                            documentId,
+                            ConvertToVSDiagnostics(diagnosticReport.Diagnostics, document, text),
+                            diagnosticReport.ResultId));
+            }
+
+            return null;
+        }
+
+        /// <summary>
+        /// Convert XamlDiagnostics to VSDiagnostics
+        /// </summary>
+        private static VSDiagnostic[]? ConvertToVSDiagnostics(XamlDiagnostic[]? xamlDiagnostics, Document document, SourceText text)
+        {
+            if (xamlDiagnostics == null)
+            {
+                return null;
+            }
+
+            var project = document.Project;
+            return xamlDiagnostics.Select(d => new VSDiagnostic()
+            {
+                Code = d.Code,
+                Message = d.Message ?? string.Empty,
+                ExpandedMessage = d.ExtendedMessage,
+                Severity = ConvertDiagnosticSeverity(d.Severity),
+                Range = ProtocolConversions.TextSpanToRange(new TextSpan(d.Offset, d.Length), text),
+                Tags = ConvertTags(d),
+                Source = d.Tool,
+                Projects = new[]
+                {
+                    new ProjectAndContext
+                    {
+                        ProjectIdentifier = project.Id.Id.ToString(),
+                        ProjectName = project.Name,
+                    },
+                },
+            }).ToArray();
+        }
+
+        private static LSP.DiagnosticSeverity ConvertDiagnosticSeverity(XamlDiagnosticSeverity severity)
+            => severity switch
+            {
+                // Hidden is translated in ConvertTags to pass along appropriate _ms tags
+                // that will hide the item in a client that knows about those tags.
+                XamlDiagnosticSeverity.Hidden => LSP.DiagnosticSeverity.Hint,
+                XamlDiagnosticSeverity.Message => LSP.DiagnosticSeverity.Information,
+                XamlDiagnosticSeverity.Warning => LSP.DiagnosticSeverity.Warning,
+                XamlDiagnosticSeverity.Error => LSP.DiagnosticSeverity.Error,
+                _ => throw ExceptionUtilities.UnexpectedValue(severity),
+            };
+
+        private static DiagnosticTag[] ConvertTags(XamlDiagnostic diagnostic)
```

Comment:
we should have some bidirectional comments so that we fixup this stuff when chanding the roslyn side (and vice-versa).

Reply:
Done. Added comments in both place.